toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+func monitorProgress(total int, results chan downloadResult, doneChannel chan struct{}) (downloadedPackages []string) {
+	downloaded := 0
+	skipped := 0
+	failed := 0
+	unavailable := 0
+	progressIncrement := 10.0
+	lastProgressUpdate := progressIncrement * -1
+
+	for done := false; !done; {
+		// Wait for a result from a worker, or the done channel to be closed (which means all workers are done)
+		select {
+		case result := <-results:
+			switch result.resultType {
+			case downloadResultTypeSkipped:
+				logger.Log.Debugf("Skipping pre-caching '%s'. File already exists", result.pkgName)
+				skipped++
+			case downloadResultTypeSuccess:
+				downloadedPackages = append(downloadedPackages, result.pkgName)
+				downloaded++
+			case downloadResultTypeFailure:
+				failed++
+			case donwloadResultTypeUnavailable:
+				logger.Log.Warnf("Pre-caching: '%s' failed, not found in any repos", result.pkgName)
+				unavailable++
+			}
+		case <-doneChannel:
+			// All workers are done, finish this iteration of the loop and then return
+			done = true
+		}
+
+		// Calculate the progress percentage and update the progress counter if needed (update every 'progressIncrement' percent)
+		completed := downloaded + skipped + failed + unavailable
+		progressPercent := (float64(completed) / float64(total)) * 100
+		if progressPercent > lastProgressUpdate+progressIncrement || done {
+			logger.Log.Infof("Pre-caching: %3d%% ( downloaded: %4d, skipped: %4d, unavailable: %4d, failed: %4d )", int(progressPercent), downloaded, skipped, unavailable, failed)
+			lastProgressUpdate = progressPercent
+		}
+	}
+	return
+}
+
+// precachePackage will attempt to download the specified package. It will return a downloadResult struct via the results.
+// This function runs with best effort, so it will return a result (of type downloadResultTypeFailure) if any error occurs
+// rather than returning an error. The caller is expected to have added to the provided wait group, while this function is
+// responsible for removing itself from the wait group. As much processing as possible is done before acquiring the
+// network operations semaphore to minimize the time spent holding it.
+func precachePackage(pkg *repocloner.RepoPackage, availablePackages map[string]string, outDir string, wg *sync.WaitGroup, results chan<- downloadResult, netOpsSemaphore chan struct{}) {
+	const (
+		downloadRetryAttempts = 2
+		downloadRetryDuration = time.Second
+	)
+
+	defer wg.Done()
+
+	// File names are of the form "<name>-<version>.<distro>.<arch>.rpm"
+	pkgName := fmt.Sprintf("%s-%s.%s.%s", pkg.Name, pkg.Version, pkg.Distribution, pkg.Architecture)
+	fileName := fmt.Sprintf("%s.rpm", pkgName)
+	fullFilePath := path.Join(outDir, fileName)
+	result := downloadResult{
+		pkgName: fileName,
+	}
+
+	// Bail out early if the file already exists
+	if exists, _ := file.PathExists(fullFilePath); exists {
```

Comment:
Why ignore the error from `PathExists`?

Reply:
This function is very unlikely to error, and I don't think we care if it does. An error here is effectively the same thing as not having the file. If there is something really wrong then we will hear about it when we try to re-write the file.